app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java

```diff
@@ -1,83 +1,311 @@
 package com.nutomic.syncthingandroid.activities;
 
-import android.Manifest;
+import android.annotation.SuppressLint;
 import android.app.Activity;
+import android.content.Context;
 import android.content.Intent;
 import android.content.SharedPreferences;
 import android.content.pm.PackageManager;
+import android.graphics.Color;
+import android.Manifest;
+import android.os.Build;
 import android.os.Bundle;
 import android.support.annotation.NonNull;
 import android.support.v4.app.ActivityCompat;
 import android.support.v4.content.ContextCompat;
+import android.support.v4.view.PagerAdapter;
+import android.support.v4.view.ViewPager;
+import android.support.v7.app.AppCompatActivity;
+import android.text.Html;
 import android.util.Log;
+import android.view.LayoutInflater;
+import android.view.MotionEvent;
 import android.view.View;
+import android.view.View.OnTouchListener;
+import android.view.ViewGroup;
+import android.view.Window;
+import android.view.WindowManager;
 import android.widget.Button;
+import android.widget.LinearLayout;
+import android.widget.TextView;
 import android.widget.Toast;
 
 import com.nutomic.syncthingandroid.R;
 import com.nutomic.syncthingandroid.SyncthingApp;
+import com.nutomic.syncthingandroid.service.Constants;
 
 import javax.inject.Inject;
 
-public class FirstStartActivity extends Activity implements Button.OnClickListener {
+public class FirstStartActivity extends Activity {
 
     private static String TAG = "FirstStartActivity";
+    private static final int REQUEST_COARSE_LOCATION = 141;
     private static final int REQUEST_WRITE_STORAGE = 142;
+    private static final int SLIDE_POS_LOCATION_PERMISSION = 1;
+
+    private ViewPager mViewPager;
+    private ViewPagerAdapter mViewPagerAdapter;
+    private LinearLayout dotsLayout;
+    private TextView[] dots;
+    private int[] layouts;
+    private Button btnBack;
+    private Button btnNext;
 
     @Inject SharedPreferences mPreferences;
 
     /**
      * Handles activity behaviour depending on {@link #isFirstStart()} and {@link #haveStoragePermission()}.
      */
+    @SuppressLint("ClickableViewAccessibility")
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         ((SyncthingApp) getApplication()).component().inject(this);
 
-        if (!isFirstStart()) {
+        /**
+         * Recheck storage permission. If it has been revoked after the user
+         * completed the welcome slides, displays the slides again.
+         */
```

Comment:
Are we going to show all slides if the permission is revoked?

Reply:
Yes, I decided to do so in the first proposal as I consider it rare that users revoke the permission if they actively use the app. (see rare one bug report on the tracker where one user did). Reason is it will never harm anything going through the slides again - also being able to grant the location permission if this was also revoked "later" after completing the wizard for the first time. Cluttering the code to handle rare cases is a maybe can do but currently don't see necessary.